[2.51.0] midx-write: fix segfault and other cleanups

derrickstolee opened the pull request:
Users have reported crashes during background maintenance. Their memory dumps appear to have an access violation in close_pack(), but it doesn't make sense based on the current implementation.

I finally figured out the problem after wading through a bunch of messiness in the midx-write.c file. This _definitely_ fixes the segfault, but it also does some of that cleanup.

This is also going upstream in gitgitgadget#1965, but hopefully this version can be fast-tracked to microsoft/git for our users who need it quickly.

derrickstolee committed:
midx-write: use cleanup when incremental midx fails

The incremental mode of writing a multi-pack-index has a few extra
conditions that could lead to failure, but these are currently
short-ciruiting with 'return -1' instead of setting the method's
'result' variable and going to the cleanup tag.

Replace these returns with gotos to avoid memory issues when exiting
early due to error conditions.

Unfortunately, these error conditions are difficult to reproduce with
test cases, which is perhaps one reason why the memory loss was not
caught by existing test cases in memory tracking modes.

Signed-off-by: Derrick Stolee <stolee@gmail.com>

derrickstolee committed:
midx-write: put failing response value back

This instance of setting the result to 1 before going to cleanup was
accidentally removed in fcb2205b77 (midx: implement support for writing
incremental MIDX chains, 2024-08-06).

Signed-off-by: Derrick Stolee <stolee@gmail.com>

derrickstolee commented:
I added a few more cleanups that I've seen in this file. I'm still looking for a smoking gun. I may reorganize these patches before sending them upstream.

derrickstolee commented:
692232b4198d4c68e6af7b77bd792fc9da99189b is the smoking gun. I'll work on an upstream version tomorrow.

@dscho @mjcheetham could you please give this a review? If you want, I can pull out everything except the "real" fix (the final commit) to avoid extra patches in the fork.

dscho commented on midx-write.c:
```diff
@@ -1143,6 +1143,7 @@ static int write_midx_internal(struct repository *r, const char *object_dir,
 		}
 	} else if (ctx.m && fill_packs_from_midx(&ctx, preferred_pack_name,
 						 flags) < 0) {
+		result = 1;
```
_sigh_. I really wish that the reviews on the Git mailing list would focus first and foremost on correctness. Stylistic concerns could come after that, if need truly be, but the correctness shouldn't be an afterthought.

derrickstolee replied:
The original patch was _huge_. Reading diffs only make this something that's easy to miss.

dscho commented on midx-write.c:
```diff
@@ -1,5 +1,3 @@
-#define DISABLE_SIGN_COMPARE_WARNINGS
```
Very nice!

For the record, before you're sending this to the Git mailing list (where reviewers tend to hyperfocus on issues that are easy to review, such as grammatical errors, over correctness issues, which are hard to review), you may want to replace "perpatrator" by "perpetrator".

derrickstolee replied:
I rewrote the message in the upstream version. Thanks.

dscho commented on midx-write.c:
```diff
@@ -24,6 +22,7 @@
 #define BITMAP_POS_UNKNOWN (~((uint32_t)0))
 #define MIDX_CHUNK_FANOUT_SIZE (sizeof(uint32_t) * 256)
 #define MIDX_CHUNK_LARGE_OFFSET_WIDTH (sizeof(uint64_t))
+#define NO_PREFERRED_PACK (~((uint32_t)0))
```
Do we need to take precautions against an actual preferred pack with the number `2^32-1`?

derrickstolee replied:
Probably not, but at least it previously would have broken at `2^31` and works a bit longer now.

dscho commented on midx-write.c:
```diff
@@ -1197,13 +1205,14 @@ static int write_midx_internal(struct repository *r, const char *object_dir,
-		struct packed_git *oldest = ctx.info[ctx.preferred_pack_idx].p;
+		struct packed_git *oldest;
 		ctx.preferred_pack_idx = 0;
+		oldest = ctx.info[ctx.preferred_pack_idx].p;
```
Wow. This is a bug fix, no? As such, shouldn't it be extracted into its own commit? Or maybe this is unintentional a change?

derrickstolee replied:
It isn't quite a bug fix. It was doing the right thing based on the incorrectly-initialized value of `preferred_pack_idx = 0`.

dscho replied:
Ah, thank you for the clarification!

dscho reviewed (dismissed):
Thank you for working on this!

I had to kick the CI build twice until it passed (and couldn't make head nor tails of the failure [because of a bug introduced in `git/git` that prevents failures to be shown in containerized builds](https://lore.kernel.org/git/e45b9487-b3ae-ed85-fd07-c92cfbf47cbb@gmx.de/)), and I have a couple of questions/suggestions, but otherwise this looks ready to go to me!

derrickstolee committed:
midx-write: only load initialized packs

The fill_packs_from_midx() method was refactored in fcb2205b77 (midx:
implement support for writing incremental MIDX chains, 2024-08-06) to
allow for preferred packfiles and incremental multi-pack-indexes.
However, this led to some conditions that can cause improperly
initialized memory in the context's list of packfiles.

The conditions caring about the preferred pack name or the incremental
flag are currently necessary to load a packfile. But the context is
still being populated with pack_info structs based on the packfile array
for the existing multi-pack-index even if prepare_midx_pack() isn't
called.

Add a new test that breaks under --stress when compiled with
SANITIZE=address. The chosen number of 100 packfiles was selected to get
the --stress output to fail about 50% of the time, while 50 packfiles
could not get a failure in most --stress runs. This test has a very
minor check at the end confirming only one packfile remaining. The
failing nature of this test actually relies on auto-GC cleaning up some
packfiles during the creation of the commits, as tests setting gc.auto
to zero make the packfile count match the number of added commits but
also avoids hitting the memory issue.

The test case is marked as EXPENSIVE not only because of the number of
packfiles it creates, but because some CI environments were reporting
errors during the test that I could not reproduce, specifically around
being unable to open the packfiles or their pack-indexes.

When it fails under SANITIZE=address, it provides the following error:

AddressSanitizer:DEADLYSIGNAL
=================================================================
==3263517==ERROR: AddressSanitizer: SEGV on unknown address 0x000000000027
==3263517==The signal is caused by a READ memory access.
==3263517==Hint: address points to the zero page.
    #0 0x562d5d82d1fb in close_pack_windows packfile.c:299
    #1 0x562d5d82d3ab in close_pack packfile.c:354
    #2 0x562d5d7bfdb4 in write_midx_internal midx-write.c:1490
    #3 0x562d5d7c7aec in midx_repack midx-write.c:1795
    #4 0x562d5d46fff6 in cmd_multi_pack_index builtin/multi-pack-index.c:305
    ...

This failure stack trace is disconnected from the real fix because it
the bad pointers are accessed later when closing the packfiles from the
context.

There are a few different aspects to this fix that are worth noting:

 1. We return to the previous behavior of fill_packs_from_midx to not
    rely on the incremental flag or existence of a preferred pack.

 2. The behavior to scan all layers of an incremental midx is kept, so
    this is not a full revert of the change.

 3. We skip allocating more room in the pack_info array if the pack
    fails prepare_midx_pack().

 4. The method has always returned 0 for success and 1 for failure, but
    the condition checking for error added a check for a negative result
    for failure, so that is now updated.

 5. The call to open_pack_index() is removed, but this is needed later
    in the case of a preferred pack. That call is moved to immediately
    before its result is needed (checking for the object count).

Signed-off-by: Derrick Stolee <stolee@gmail.com>

derrickstolee committed:
midx-write: use uint32_t for preferred_pack_idx

midx-write.c has the DISABLE_SIGN_COMPARE_WARNINGS macro defined for a
few reasons, but the biggest one is the use of a signed
preferred_pack_idx member inside the write_midx_context struct. The code
currently uses -1 to indicate an unset preferred pack but pack int ids
are normally handled as uint32_t. There are also a few loops that search
for the preferred pack by name and those iterators will need updates to
uint32_t in the next change.

For now, replace the use of -1 with a 'NO_PREFERRED_PACK' macro and an
equality check. The macro stores the max value of a uint32_t, so we
cannot store a preferred pack that appears last in a list of 2^32 total
packs, but that's expected to be unreasonable already. This improves the
range from 2^31 already.

There are some careful things to worry about with initializing the
preferred pack in the struct and using that value when searching for a
preferred pack that was already incorrect but accidentally working when
the index was initialized to zero.

Signed-off-by: Derrick Stolee <stolee@gmail.com>

derrickstolee committed:
midx-write: reenable signed comparison errors

Remove the remaining signed comparison warnings in midx-write.c so that
they can be enforced as errors in the future. After the previous change,
the remaining errors are due to iterator variables named 'i'.

The strategy here involves defining the variable within the for loop
syntax to make sure we use the appropriate bitness for the loop
sentinel. This matters in at least one method where the variable was
compared to uint32_t in some loops and size_t in others.

While adjusting these loops, there were some where the loop boundary was
checking against a uint32_t value _plus one_. These were replaced with
non-strict comparisons, but also the value is checked to not be
UINT32_MAX. Since the value is the number of incremental multi-pack-
indexes, this is not a meaningful restriction. The new die() is about
defensive programming more than it being realistically possible.

Signed-off-by: Derrick Stolee <stolee@gmail.com>

derrickstolee commented:
Making note that I have gitgitgadget#1965 prepared for upstream. It includes a test case that verifies this fix for the memory bug.

derrickstolee commented:
Sorry to cause noise with the new version, but this matches the GGG version to minimize differences in future integration things.

dscho commented:
> this matches the GGG version to minimize differences in future integration things.

Much appreciated!
